### PR TITLE
Use Parallel::ForkManager in place of shell syntax

### PR DESCRIPTION
**Pull request by dkj**
Clean up and update with current functionality existing branch for this Parallel::ForkManager use.

Should avoid corrupt cram and crai we're seeing.

### CODE REVIEW

File: lib/npg_common/sequence/BAM_MarkDuplicate.pm [hunk 1]
```diff
@@ -393,6 +397,609 @@ has 'replace_file'=> (isa        => 'Bool',
...
+       }
+     }
+
+     if( ! $self->not_strip_bam_tag() ){
+         $mark_duplicate_cmd = $self->bam_tag_stripper_cmd();
+     }
+
+  }else{
+     $mark_duplicate_cmd = $self->mark_duplicate_cmd();
+     my $using_pipe;
```
Comment (mgcam) [inline lib/npg_common/sequence/BAM_MarkDuplicate.pm]: $using_pipe variable - is it needed?

**Comment by mgcam** [comment]
A list of fifos is created independently of the list of commands. This potentially can lead to a mismatch in the two lists. It would be better to create these two lists simultaneously. An option of a hash with a short command names as keys and then a full commad and fifo name might be better. I am not sure what is the value of attributes for fifo names. All of them are created in a similar way. It might be sufficient to have  (1) a hash where the keys are short command names and values are commands themselves and (2) a function that, given a short command name returns a full fifo path.

**Comment by mgcam** [comment]
so lines 441-451 will be just a loop over available short command names, no additional logic.

**Comment by dkj** [comment]
Re cleaning up the logic for fifo and command generation:
1. it is currently messy
2. I'm not clear on the hash suggestion (fifos and commands to fork are not one to one  - a lot of the fifo and command generation pattern is the same but there are a few variants)
3. I think putting fifo generation in the same logic for the initial command it is used in would
  1. be clearer
  2. require more rewrite and testing than I'm keen on for now
     * I'd like such a big cleanup to be a move to P4, either the BAM_MarkDup module functionality or the whole BAM_Alignment

Only suggesting this patch/PR as a fix for current production breakage.

**Comment by mgcam** [comment]
On 14/02/2015 20:42, dkj wrote:
> Only suggesting this patch/PR as a fix for current production breakage.
OK, understand. I'll do a local merge and run the tests.



-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Pull request merged**